core/amber/src/test/scala/edu/uci/ics/texera/workflow/operators/specializedFilter/SpecializedFilterOpExecSpec.scala

```diff
@@ -0,0 +1,226 @@
+        val empty = Iterator()
+        if (result == empty) 0 else 1
+      })
+      .reduce((count1, count2) => count1 + count2)
+
+    assert(resultSize == 0)
+    opExec.close()
+  }
+
+  it should "throw NullPointerException when predicates is null" in {
+    val opExec = new SpecializedFilterOpExec(new SpecializedFilterOpDesc())
+    opExec.open()
+    val linkID1 = linkID()
+
+    assertThrows[NullPointerException] {
+      Range(0, 4).map(i => {
+        opExec.processTexeraTuple(Left(allNullTuple()), linkID1)
+      })
+    }
+
+    opExec.close()
+  }
+
+  it should "do nothing when predicates is an empty list" in {
+    val opExec = new SpecializedFilterOpExec(new SpecializedFilterOpDesc(), asList())
+    opExec.open()
+    val linkID1 = linkID()
+
+    val range = Range(0, 4)
+    val resultSize = range
+      .map(_ => {
+        val result = opExec.processTexeraTuple(Left(allNullTuple()), linkID1)
+        val empty = Iterator()
+        if (result == empty) 0 else 1
+      })
+      .reduce((count1, count2) => count1 + count2)
+
+    assert(resultSize == 0)
+    opExec.close()
+  }
+
+  it should "not have null comparisons be affected by values, only comparison types" in {
+    val predicates = asList(new FilterPredicate("string", ComparisonType.IS_NULL, markerValue()))
+    val opExec = new SpecializedFilterOpExec(new SpecializedFilterOpDesc(), predicates)
+    opExec.open()
+
+    val range = Range(0, 7)
+    val linkID1 = linkID()
+
+    val resultSize = range
+      .map(_ => {
+        val result = opExec.processTexeraTuple(Left(allNullTuple()), linkID1)
+        val empty = Iterator()
+        if (result == empty) 0 else 1
+      })
+      .reduce((count1, count2) => count1 + count2)
+
+    assert(resultSize == range.size)
+    opExec.close()
+  }
+
+  it should "filter the right values with 2 input streams with is null" in {
+    val predicates = asList(
+      new FilterPredicate("bool", ComparisonType.IS_NULL, null)
+    )
+    val opExec = new SpecializedFilterOpExec(new SpecializedFilterOpDesc(), predicates)
+    opExec.open()
+
+    val inclusionRange = Range(0, 7)
+    val exclusionRange = Range(0, 9)
+    val linkID1 = linkID()
+    val linkID2 = linkID()
+    var countIncluded = 0
+
+    inclusionRange.map(_ => {
+      val result = opExec.processTexeraTuple(Left(allNullTuple()), linkID1)
+      countIncluded = if (result == Iterator()) countIncluded else countIncluded + 1
+      result
+    })
+    assert(opExec.processTexeraTuple(Right(InputExhausted()), linkID1).isEmpty)
+
+    exclusionRange.map(_ => {
+      val result = opExec.processTexeraTuple(Left(nonNullTuple()), linkID2)
+      countIncluded = if (result == Iterator()) countIncluded else countIncluded + 1
+      result
+    })
+
+    assert(countIncluded == inclusionRange.size)
+
+    opExec.close()
+  }
+
+  it should "filter the right values with 2 input streams with is not null" in {
```

Comment:
A basic question. Why are we considering two input streams for filter?

Reply:
I think these cases can be reduced to the remaining two cases:
1. it should "filter out non-null tuples when filter is is_null"
2. it should "output non-null tuples when filter is is_not_null"

Reply:
Yes I agree, I initially thought that all operators should support multi-stream input, but it makes sense now that filter has only a single data source. Sure I can make the corresponding changes.